state/cache.go

```diff
@@ -0,0 +1,62 @@
+package state
+
+//go:generate mockgen -source cache.go -destination cache_mocks.go -package state
+import (
+	"sync"
+
+	cc "github.com/Fantom-foundation/Carmen/go/common"
+	"github.com/ethereum/go-ethereum/common"
+)
+
+// CodeCache serves a cache for hashed address code.
+type CodeCache interface {
+	// Get returns code hash for given addr and code.
+	Get(addr common.Address, code []byte) common.Hash
+}
+
+// codeKey represents the key for CodeCache.
+type codeKey struct {
+	addr common.Address
+	code string
+}
+
+// NewCodeCache creates new instance of CodeCache that stores already retrieved code hashes.
+func NewCodeCache(capacity int) CodeCache {
+	if capacity <= 0 {
+		return &codeCache{}
+	}
+	return &codeCache{
+		cache: cc.NewLruCache[codeKey, common.Hash](capacity),
+		mutex: sync.Mutex{},
+	}
+}
+
+type codeCache struct {
+	cache cc.Cache[codeKey, common.Hash]
+	mutex sync.Mutex
+}
+
+// Get returns code hash for given addr and code.
+// If hash does not exist within the cache, it is created and stored.
+// This operation is thread-safe.
+func (c *codeCache) Get(addr common.Address, code []byte) common.Hash {
+	k := codeKey{addr: addr, code: string(code)}
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+	h, exists := c.cache.Get(k)
```

Comment:
This lookup is quite expensive. The code needs to be copied (one full pass), hashed for the cache lookup (one full pass), and equality compared for the hash look up (at least one full pass). The actual hashing only requires a single pass and no memory allocation.

You should micro-benchmark this function and compare it to the `createCodeHash` function to see whether even in best-case scenarios a lookup is faster than re-hashing.